### PR TITLE
Add support for Cardano cryptocurrency

### PR DESCRIPTION
**Pull request by detrandix**

### CODE REVIEW

File: lib/coins/ada.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+const b58 = require('b58')
+const cbor = require('cbor-sync')
+const CRC32 = require('crc-32'); 
+
+module.exports = {depositUrl, parseUrl}
+
+function parseUrl(network, address) {
+  console.log('DEBUG16: [%s] *%s*', network, address)
```
Comment (rizip1) [inline lib/coins/ada.js]: I suppose this should be removed.

File: lib/coins/ada.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  // Cardano in Byron era not recognize network type (main/test) by address prefixes like f.e. bitcoin
```
Comment (rizip1) [inline lib/coins/ada.js]: Typo: `not` -> `does not`, `f.e.` -> `e.g.`

File: lib/coins/ada.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  if (!validate(address)) return null
```
Comment (rizip1) [inline lib/coins/ada.js]: Convention: Lets move `validate` above this `parseUrl` function, so it is not used sooner than defined.

Comment (rizip1) [inline lib/coins/ada.js]: This is how other currencies handle nonvalid case right? No need to throw some exception? (just asking)

File: lib/coins/ada.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
+const b58 = require('b58')
```
Comment (rizip1) [inline lib/coins/ada.js]: This package has 8 downloads in a week so I am quite afraid of using it. However, I believe there should be something like that already available as other cryptocurrencies should need it. Could you please check current dependencies or how other currencies do this?

Comment (rizip1) [reply lib/coins/ada.js]: It seems you can use already install "base-x" package.

File: lib/coins/ada.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+  return address
+}
+
+function depositUrl(address, amount) {
+  return `${address}?amount=${amount}`
+}
+
+function validate(address) { 
+  const bytes = b58.decode(address)
```
Comment (rizip1) [inline lib/coins/ada.js]: What happens if an address is not in base58? Would not it throw some exception?

File: test/unit/test-coin-parse-url.js [hunk 6]
```diff
@@ -64,3 +65,27 @@ test('Should throw for invalid ZEC address', t => {
...
+    'Ae2tdPwUPEZKmwoy3AU3cXb5Chnasj6mvVNxV1H11997q3VW5ihbSfQwGpm' // Yoroi like address
+  ]
+  
+  for (let addr of addresses) {
+    let parsed = parseADAAddress('main', addr)
+    t.is(parsed, addr)
+  }
+})
+
+test('Should recognize invalid ADA address', t => {
```
Comment (rizip1) [inline test/unit/test-coin-parse-url.js]: What about test where address is not base58 encoded?

Comment (rizip1) [inline test/unit/test-coin-parse-url.js]: I would either rename it to "Should recognize invaliad ADA addresses" as you test multiple addresses, but rather would create a separate test for each "invalid" case.

File: test/unit/test-coin-parse-url.js [hunk 7]
```diff
@@ -64,3 +65,27 @@ test('Should throw for invalid ZEC address', t => {
   } catch (error) { }
   t.true(spy.threw())
 })
+
+test('Should parse ADA address (Byron era)', t => {
```
Comment (rizip1) [inline test/unit/test-coin-parse-url.js]: Also here it could be better to divide this into two standalone tests.

File: lib/coins/ada.js (hunk 3 above)
Comment (detrandix) [reply lib/coins/ada.js]: Yes, this is how non-valid cases are handled.

File: lib/coins/ada.js (hunk 4 above)
Comment (detrandix) [reply lib/coins/ada.js]: There is also a `bs58check` package, but that does BTC (and other similar currencies) specific CRC checking (Cardano have different algorithm), so from this point of view `base-x` looks good.

File: test/unit/test-coin-parse-url.js (hunk 6 above)
Comment (detrandix) [reply test/unit/test-coin-parse-url.js]: OK

File: lib/coins/ada.js (hunk 5 above)
Comment (detrandix) [reply lib/coins/ada.js]: I will add try/catch blocks and return false

File: test/unit/test-coin-parse-url.js (hunk 6 above)
Comment (detrandix) [reply test/unit/test-coin-parse-url.js]: OK, will do separate tests :)

File: lib/coins/ada.js (hunk 1 above)
Comment (detrandix) [reply lib/coins/ada.js]: This log is in all other coins (except Ethereum) - so are you sure about remove this?

File: lib/coins/ada.js (hunk 3 above)
Comment (detrandix) [reply lib/coins/ada.js]: You are right, they fire exceptions, but for BTC this is little hidden in bs58check library

File: lib/coins/ada.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
+const BASE58 = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'
```
Comment (rizip1) [inline lib/coins/ada.js]: It seems you miss `I` here, is not there a better way to create that? Using some JS methods returning chars for ascii positions?

File: lib/coins/ada.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+const b58 = require('base-x')(BASE58)
+const cbor = require('cbor-sync')
+const CRC32 = require('crc-32'); 
+
+module.exports = {depositUrl, parseUrl}
```
Comment (rizip1) [inline lib/coins/ada.js]: Please move this at the bottom of file

File: lib/coins/ada.js (hunk 1 above)
Comment (rizip1) [reply lib/coins/ada.js]: Ok in that case leave it here

File: lib/coins/ada.js (hunk 9 above)
Comment (detrandix) [reply lib/coins/ada.js]: They use this in all other coins files, so I pick this as a convention, so do you think is a good idea to move that?

File: lib/coins/ada.js (hunk 8 above)
Comment (detrandix) [reply lib/coins/ada.js]: > Compared with Base64, the following similar-looking letters are omitted: 0 (zero), O (capital o), I (capital i) and l (lower case L) as well as the non-alphanumeric characters + (plus) and / (slash).

File: lib/coins/ada.js (hunk 9 above)
Comment (rizip1) [reply lib/coins/ada.js]: Ok then leave it at the top

File: lib/coins/ada.js (hunk 8 above)
Comment (rizip1) [reply lib/coins/ada.js]: sorry, my bad

File: lib/coins/ada.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  const addrData = cborData[0]
+  const addrChecksum = cborData[1]
+  
+  const checksum = CRC32.buf(addrData) >>> 0 // force usigned integer
+  
+  if (addrChecksum !== checksum) {
+    throw new Error(`Ivalid address ${address}: invalid checksum`)
+  }
+
+  return true
```
Comment (rizip1) [inline lib/coins/ada.js]: This return statement now does not make much sense as there are exceptions

**Pull request closed without merging**